File: src/main/java/io/github/marmer/GreetingProvider.java
package io.github.marmer;

import java.time.LocalTime;

public class GreetingProvider {

    public static final LocalTime MORNING_START = LocalTime.of(6, 0);
    public static final LocalTime NOON_START = LocalTime.of(12, 0);
    public static final LocalTime NIGHT_START = LocalTime.of(20, 0);
    private final SystemTimeProvider systemTimeProvider;

    public GreetingProvider(final SystemTimeProvider systemTimeProvider) {
        this.systemTimeProvider = systemTimeProvider;
    }

    public String getGreetingFor(final String name) {
        final LocalTime now = now();
        if (isMorning(now)) {
            return "Buenos Dias " + name;
        } else if (isNoon(now)) {
            return "Buenos Tarde " + name;
        } else {
            return "Buenos Noche " + name;
        }
    }

    private boolean isNoon(final LocalTime now) {
        return isBetween(now, NOON_START, NIGHT_START);
    }

    private boolean isMorning(final LocalTime now) {
        return isBetween(now, MORNING_START, NOON_START);
    }

    private boolean isBetween(final LocalTime now, final LocalTime startInclusive, final LocalTime endExclusive) {
        return (now.equals(startInclusive) || now.isAfter(startInclusive)) &&
                now.isBefore(endExclusive);
    }

    private LocalTime now() {
        return systemTimeProvider.getNow();
    }
}

File: src/main/kotlin/io/github/marmer/GreetingProvider.kt
package io.github.marmer

import java.time.LocalTime

class GreetingProvider(private val systemTimeProvider: SystemTimeProvider) {
    fun greetingFor(name: String): String {
        val now = systemTimeProvider.currentTime()
        return when {
            now.isMorning() -> "Buenos dias $name"
            now.isEvening() -> "Buenos tarde $name"
            else -> "Buenos noche $name"
        }
    }

    private fun LocalTime.isMorning(): Boolean =
        this isEqualOrAfter morningStart &&
                this isBefore eveningStart

    private fun LocalTime.isEvening(): Boolean =
        this isEqualOrAfter eveningStart &&
                this isBefore nightStart

    private infix fun LocalTime.isEqualOrAfter(time: LocalTime): Boolean = equals(time) || isAfter(time)
    private infix fun LocalTime.isBefore(time: LocalTime): Boolean = isBefore(time)

    private val morningStart = LocalTime.of(6, 0)
    private val eveningStart = LocalTime.of(12, 0)
    private val nightStart = LocalTime.of(20, 0)
}

